tchannel/tornado/hyperbahn.py

```diff
@@ -20,90 +20,140 @@
+
+        Returns the result of the first ad request.
+        """
+        if self.running:
+            raise Exception('Advertiser is already running')
+        if self.io_loop is None:
+            self.io_loop = tornado.ioloop.IOLoop.current()
+
+        self.running = True
+        answer = tornado.gen.Future()
+        self._schedule_ad(0, answer)
+        return answer
+
+    def stop(self):
+        self.running = False
+        if self._next_ad is not None:
+            t, self._next_ad = self._next_ad, None
+            self.io_loop.remove_timeout(t)
+
+    def _schedule_ad(self, delay=None, response_future=None):
+        """Schedules an ``ad`` request.
+
+        :param delay:
+            Time in seconds to wait before making the ``ad`` request. Defaults
+            to self.interval_secs. Regardless of value, a jitter of
+            self.interval_max_jitter_secs is applied to this.
+        :param response_future:
+            If non-None, the result of the advertise request is filled into
+            this future.
+        """
+        if not self.running:
+            return
+
+        if delay is None:
+            delay = self.interval_secs
+
+        delay += random.uniform(0, self.interval_max_jitter_secs)
+        self._next_ad = self.io_loop.call_later(delay, self._ad,
+                                                response_future)
+
+    @tornado.gen.coroutine
+    def _ad(self, response_future=None):
+        self._next_ad = None
+        try:
+            response = yield self.tchannel.request(service='hyperbahn').send(
+                arg1='ad',  # advertise
+                arg2='',
+                arg3=json.dumps({
+                    'services': [
+                        {
+                            'serviceName': self.service,
+                            'cost': 0,
+                        }
+                    ]
+                }),
+                headers={'as': 'json'},
+                retry_limit=0,
+                ttl=self.ttl_secs,
             )
-
-        response = yield _advertise(tchannel, service)
-
-        if response is not None and response.code is StatusCode.ok:
-            break
-
-        attempt_counter += 1
-        attempt_counter, delay_time = _prepare_next_ad(attempt_counter)
-
-        yield tornado.gen.sleep(delay_time)
-
-    raise tornado.gen.Return(response)
+        except Exception as e:
+            log.info('Failed to register with Hyperbahn: %s', e, exc_info=True)
+            if response_future is not None:
+                response_future.set_exc_info(sys.exc_info())
+        else:
+            if response.code != StatusCode.ok:
+                log.info('Failed to register with Hyperbahn: %s', response)
+            else:
+                log.info('Successfully registered with Hyperbahn')
+            if response_future is not None:
+                response_future.set_result(response)
+        finally:
+            self.io_loop.spawn_callback(self._schedule_ad)
```

Comment:
Yeah, but this is slightly cheaper since it doesn't associate the stack context with the call. Made sense given that this is a background loop.